rfcs/2020-04-20-Optimizer-minimize.md

```diff
@@ -0,0 +1,293 @@
+
+Despite the complexity of `_minimize`, it covers only a small subset of possible optimization logic. Keras continues to receive valid requests to support more custom optimization logic (including adding hooks for different aggregation methods, different methods of loss reduction, etc). To continue expanding support for these items, Keras needs to rely on a unified API that keeps `Optimizer` implementation details from leaking into the main training loop code.
+
+The proposal below shows how this can be accomplished, and the examples section shows how this can be applied to 3 use cases: gradient clipping, mixed precision, and `Horovod`.
+
+### Custom training loops:
+
+The logic above also applies to custom training loops. The design should allow custom training loops to be written so that they work with any `Optimizer`.
+
+
+## User Benefit
+
+This design will allow users to write full-featured training loops that work for all `Optimizer`s. This design will allow users to easily perform custom gradient clipping and other transformations.
+
+## Design Proposal
+
+`Optimizer` class:
+
+```python
+class Optimizer(object):
+  def __init__(self,
+               transform_gradients=None,
+               aggregate_gradients=all_reduce_sum):
+     self.aggregate_gradients_fn = aggregate_gradients
+     self.transform_gradients_fns = transform_gradients_fns
+
+  def _transform_loss(self, loss):
+    # Can be overridden in subclasses
+    return loss
+
+  def _get_gradients(self, loss, variables, tape):
+    # Can be overridden to use jacobian, etc.
+    return tape.gradient(loss, variables)
+
+  def _transform_unaggregated_gradients(self, grads_and_vars):
+    # Can be overridden in subclasses
+    return grads_and_vars
+
+  def _aggregate_gradients(self, grads_and_vars):
+    # Can still be overridden in subclasses if needed
+    if self.aggregate_gradients_fn:
+      grads_and_vars = self.aggregate_gradients_fn(
+         grads_and_vars)
+    return grads_and_vars
+
+  def _transform_gradients(self, grads_and_vars):
+    # Can still be overridden in subclasses if needed
+    if self.transform_gradients_fns:
+      for fn in self.transform_gradients_fns:
+        grads_and_vars = fn(grads_and_vars)
+    return grads_and_vars
+   
+  def _apply_updates(self, distribution, grads_and_vars, ...):
+    # Calls _resource_apply_{dense | sparse}
+    # Variable updating math is still in _resource_apply_{dense | sparse}
+  
+  def minimize(self, loss, variables, tape=None):
+    grads_and_vars = self.compute_gradients(loss, variables, tape)
+    self.apply_gradients(grads_and_vars)
+
+  def compute_gradients(
+      self,
+      loss,
+      variables,
+      tape=None,
+      all_reduce_sum_gradients=False):
```

Comment:
why is the arg named `all_reduce_sum_gradients` even though users may have changed the aggregation from the default?

Reply:
backwards compatibility I think

Reply:
Fixed the names, meant to have the same name as our existing param, which shipped in 2.2 as `experimental_aggregate_gradients`